### ISSUE TITLE
LFO does not work like a typical LFO

### ISSUE THREAD

**Issue by bryc** [greatest-ape/OctaSine#40]
In most synths, as the amount is increased, it is applied in both directions equally and linearly. In OctaSine, it only goes in one direction at a time, and it appears to increase exponentially.

I feel this is going to throw people off; I've personally never encountered an LFO working like this before.

A vibrato or tremolo, or other traditional LFO effect requires the same amount to be applied in both directions, and ideally is linear, for less precision needed during modwheel performances.

I can't make a satisfying vibrato using OctaSine's LFOs, it seems more geared toward atonal or effect sounds, which is a bit limiting.

It would be good if each LFO had a mode switch, to go from the current LFO (assuming you wish to keep it at all) to a more traditional one. Also, I'd suggest adding Sine to the waveform list. A slowly modulated triangle does not achieve the same range and shape as a sine at the crest point, it can have a significant effect on the sound.

Also, the pitch bend doesn't seem to react, so that cannot be used to get around this.

**Comment by greatest-ape** [comment]
@bryc Those are all good points.

What do you think of this?

[OctaSine-tmp-2022-02-03-Windows.zip](https://github.com/greatest-ape/OctaSine/files/7993816/OctaSine-tmp-2022-02-03-Windows.zip)
[OctaSine-tmp-2022-02-03-macOS-Intel.zip](https://github.com/greatest-ape/OctaSine/files/7993817/OctaSine-tmp-2022-02-03-macOS-Intel.zip)

**Comment by bryc** [comment]
Sorry for the late reply, having a look now!

First, I loaded up my old project, and my presets now seem very broken with this one. The envelope attack rate is much slower and feels less punchy. If you didn't change any of that, it could be a new problem.

As for the LFO, I can now hear it going in both directions. But the LFO depth still sounds exponential instead of linear. The extremes of the waveform are getting 'tugged', which betray the actual wave shape, and almost induce odd rhymical patterns. So it's not quite there yet.

Did a quick comparison video:

https://user-images.githubusercontent.com/1408749/152779908-7f6adb8b-5198-4798-9604-6e756d99b7c2.mp4

**Comment by greatest-ape** [comment]
@bryc Thanks for trying out the changes!

Quite a few things were changed in version 0.6, including things that can affect how patches sound, so if you used an older version, you might have been affected. Also, since I'm completely reworking LFOs for this issue, old patches will risk will sound differently. It's pretty difficult (and in some cases impossible) to attempt to make patches sound the same as before when changing how audio is generated, unfortunately.

There were some envelope changes. If you think the attacks are too slow even when made as short as possible, I will consider making them faster. It's a tradeoff with avoiding audio artefacts.

I've done some more LFO changes. I found out why they were not affecting values in a linear manner. Now, when targeting volume and frequency parameters, the effect should be linear. Parameters like mod index, however, will continue to scale non-linearly in the same manner as when turning the knobs. There is also a new LFO shape picker displaying the waveform.

It would be great if you could try out the changes once again :-)

[OctaSine-tmp-2022-02-21-Windows.zip](https://github.com/greatest-ape/OctaSine/files/8111027/OctaSine-tmp-2022-02-21-Windows.zip)
[OctaSine-tmp-2022-02-21-macOS-Intel.zip](https://github.com/greatest-ape/OctaSine/files/8111030/OctaSine-tmp-2022-02-21-macOS-Intel.zip)

**Comment by bryc** [comment]
> Quite a few things were changed in version 0.6, including things that can affect how patches sound, so if you used an older version, you might have been affected.

What I meant was, I did use 0.6 in my project, and when I tested [OctaSine-tmp-2022-02-03-Windows.zip](https://github.com/greatest-ape/OctaSine/files/7993816/OctaSine-tmp-2022-02-03-Windows.zip), it caused the issues I described when I loaded that same project.

Oddly enough, [OctaSine-tmp-2022-02-21](https://github.com/greatest-ape/OctaSine/files/8111027/OctaSine-tmp-2022-02-21-Windows.zip) doesn't seem to suffer from the same envelope attack issue in the previous test version. Instead, I roughly get back to where I was by altering the Mod Indexes.

Anyway, the LFOs now behave much more how one would expect - nice work!

Made a happy little tune using sine LFOs on volume/pitch/panning/modulation - basically ran out of LFOs 😊 
https://user-images.githubusercontent.com/1408749/155082204-3ac14504-ccdf-4f22-8c76-b154ee228eb6.mp4
